File: leetcode/cpp/LinkedListCycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        // IMPORTANT: Please reset any member data you declared, as
        // the same Solution instance will be reused for each test case.
        if (! head || ! head->next) return false;
        ListNode *slow = head, *fast = head->next;
        
        while (fast->next && fast->next->next) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) return true;
        }
        
        return false;
    }
};

/*

Linked List Cycle
Total Accepted: 2253 Total Submissions: 6002

Given a linked list, determine if it has a cycle in it.

Follow up:
Can you solve it without using extra space? 
 */

File: leetcode_java/LinkedListCycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null) return false;
        
        ListNode fast = head.next, slow = head;
        
        while (fast != null && fast.next != null) {
            if (fast == slow) return true; // it works too if put this at the end of cycle.
            slow = slow.next;
            fast = fast.next.next;
        }
        
        return false;
    }
}

/**
Linked List Cycle
Difficulty: Medium

Given a linked list, determine if it has a cycle in it.

Follow up:
Can you solve it without using extra space? 
 */
